### DISCUSSION TITLE
Making a Pendant, MPG, Keypad controller

### DISCUSSION THREAD

**Discussion by stevempotter** [phil-barrett/grblHAL-teensy-4.x#27] [Q&A]
I have a grandiose plan to build a wireless pendant for my CNC, using two ESP32 microprocessors, and [ESP-NOW low-latency communication](https://randomnerdtutorials.com/esp-now-esp32-arduino-ide/). I have all the hardware and am embarking on the programming side of things, specifically, interfacing to Phil's Teensy breakout board. Phil has pointed me to some of the existing code (keypad and MPG plugins), which unfortunately are in Terje's C code, which is not very well explained or commented, and way past my level of C coding skills. I am programming in C++ in Arduino IDE. So if there are others out there who can help me get the coding side of things sorted, please post here!

I will use the I2C header on the Teensy Breakout Board (BoB), which has a strobe input.
**Here is my first question:** What is the Teensy controller expecting me to do with the strobe line on the I2C bus?

If anyone has built any sort of keypad or manual pulse generator for jogging their machine, please let us know what you did in this thread. Here is my prototype at the moment:

![Potter Pendant Prototype](https://user-images.githubusercontent.com/5586932/142766506-b531e2df-0271-4b25-8d61-290f9de09bd0.jpg)

**Comment by phil-barrett** [comment]
Well, the best build (by far) and first was done by @terjeio himself!  [Available here.](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack)
![IMG_8515](https://user-images.githubusercontent.com/16570958/142776193-c58107ae-ad83-4014-a39d-83ca1a462abf.jpg)
(Terje, sorry for copying your image, I didn't see a way to url link images in markdown)

**Comment by terjeio** [comment]
There are some tinkering with implementing pendants/offline controllers. See [this](https://github.com/grblHAL/core/issues/80) and [this](https://github.com/grblHAL/core/discussions/82) issue for examples.

The [I2C keypad plugin](https://github.com/grblHAL/Plugin_I2C_keypad) could be a starting point for I2C communication - it uses single character commands for jogging etc. Recently I also added support for serial input to that. If using I2C the strobe pin is for telling the plugin that a command character is available, this is done by pulling it low. For jogging it should be held low for the duration of a keypress, for other commands only briefly - 10 microseconds or so. When pulled low the plugin will request the keycode from the pendant. 

[Documentation](http://svn.io-engineering.com/grblHAL/html/hal_8h.html) is a large project in itself - I have started writing some and for now I just follow up those who want to explore coding with grblHAL in discussions such as those linked above. Feel free to do so yourself.

Phil, right-clicking an image gives you the option of copying the URL, put a ! (pling?) in front of the link to embed it:

`![img](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack/raw/master/media/IMG_8515.jpg)`

-> 

![img](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack/raw/master/media/IMG_8515.jpg)

**Reply by stevempotter** [reply]
@terjeio, your lathe pendant is beautiful and obviously, you put a lot of work into making it and the programming of it. Unfortunately, I have only passing familiarity with bit math, pointers, structs, and the various shortcuts you like to use in programming in C (like  the |= operator). I have been coding since 1984 in BASIC, Pascal, C++ (Arduino IDE), and LabView, but not C. To help make your code more usable and extendable by others, I encourage you to please do more commenting and documentation to explain how the code works, what it does and how to use it.

My first question: how and where do you plug in a plugin?

**Reply by terjeio** [reply]
My plan is to go through all the code and add [documentation](http://svn.io-engineering.com/grblHAL/html/hal_8h.html) that can be extracted by Doxywizard. This will take time... 

For how to write plugins there are examples in the [Templates](https://github.com/grblHAL/Templates/blob/master/my_plugin/probe%20select/my_plugin.c) repo and in the _Plugin_*  repositories - with some comments. Tutorials should be written as well, but currently the project is mainly maintained by me and I simply do not have time for all that should be done.
Luckily some of the collaborators has indicated that they will do write-ups of what they have done - that will surely help.

> My first question: how and where do you plug in a plugin?

For the plugins that are in separate repositories it is a matter of enabling them in [my_machine.h](https://github.com/grblHAL/iMXRT1062/blob/110dce4c92861d3fd898094da4d1bf0e9bb338d1/grblHAL_Teensy4/src/my_machine.h#L56-L72). Most of these will now report any pins they use in the `$pins` report.

If it is your own plugin it has to be placed in the same folder as [driver.c](https://github.com/grblHAL/iMXRT1062/tree/master/grblHAL_Teensy4/src) and contain a function `void my_plugin_init (void)  {...}`. This will be called on startup and is where it "hooks" itself into the controller code, typically by modifying function pointers and/or registering callback functions. A lot of stuff can be done this way, and this without modifying a single line of code elsewhere. See the Templates examples.

Again, feel free to [start a discussion](https://github.com/grblHAL/core/discussions) over at the grblHAL repo - it will more visible for other users and will also give me pointers on where to start documenting.

**Comment by stevempotter** [comment]
Espressif recently made a new video describing MANY new features of ESP-NOW. It is even better than it already was. Have a look, and I bet you will agree that this is ideal for a wireless pendant application:
https://youtu.be/QmvMtgNs9r8

**Reply by terjeio** [reply]
A plugin for grblHAL that handles the ESP-NOW transfer protocol and a grblHAL specific protocol on top of that has to be written. I can provide guidance for how to write the plugin. The [Bluetooth plugin](https://github.com/grblHAL/Plugins_Bluetooth) could be used as a template?

**Reply by stevempotter** [reply]
I believe an ESP-NOW plugin only needs to be written when grblHAL is being run on the receiver ESP32. For now, I will continue to run grblHAL on the Teensy 4.1, and have the receiver ESP32 talk to it via its I2C line and strobe. 

It would be helpful to see all the characters exchanged between a keypad or encoder and grblHAL, via I2C. Can you post a typical conversation here, with commenting to explain things?  As I have said before, I can't understand the keypad C code, or the keypad plugin code. But if I knew exactly what my Arduino C++ sketch was supposed to be saying to the Teensy (and what to expect from the Teensy), I think I can proceed to get the wireless pendant working.

**Reply by terjeio** [reply]
It is quite simple, when the strobe signal is pulled low a byte is requested from ESP32 (I2C address 0x49) which has to respond with the command character. I set up an interrupt handler in my code to respond to the request. Jog command characters are a bit special in that the strobe signal has to be held low while the key is pressed, other keys should just send a short pulse to trigger the request.

I have just received my [Pi Pico MPG & DRO controller PCB](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack/tree/master/Pi_Pico_Board) so I will be working on this the coming days, my plan is to add MPG mode support to the Teensy driver and validate use of the serial port instead of I2C for command character transfer. MPG mode allows the pendant to take full control, it then becomes a sender. This allows more advanced functionality...

**Reply by stevempotter** [reply]
>the command character

Which character is that? 
Or did you mean, "The sequence of characters, such as $J=X20F3000"?
If so, more than one byte needs to be requested. How is that done?

That is good news that you are forging ahead on the Pi Pico! That MPG mode is something I would like to use.
Which serial port? USB host? USB device? Micro-USB? To me I2C is a serial port, too. Can the Teensy presently accept commands into its micro-USB port when it is talking to ioSender via its ethernet port?

**Reply by terjeio** [reply]
> Which character is that?

[These](https://github.com/grblHAL/Plugin_I2C_keypad/blob/27f34a469d29b31e715c418c13314361a0b2381c/keypad.c#L172-L263), some are defined [here](https://github.com/grblHAL/Plugin_I2C_keypad/blob/27f34a469d29b31e715c418c13314361a0b2381c/keypad.h#L36-L49) and [here](https://github.com/grblHAL/core/blob/af29d13c2b67b702f0331a24c4f3fb1e6f1ceed4/grbl.h#L98-L141).

Edit: fixed incorrect link.

> Or did you mean, "The sequence of characters, such as $J=X20F3000"?

No, the keypad plugin only accepts single character commands.

> Which serial port?

UART, and sadly there is only one available. Using I2C for the full grbl protocol may be possible but I am not going to implement that - you can do that yourself if you want to. USB needs code support but there is a slight problem with USB ports - there is no good way to determine if a client is connected or not, if sending data to a non-existing client buffers will fill up and freeze the system.

> Can the Teensy presently accept commands into its micro-USB port when it is talking to ioSender via its ethernet port?

Yes, with code support - I may add an option for redirecting the input to the keypad plugin.
FYI I am not going to allow two active streams sending gcode at the same time, the MPG protocol uses a handshake signal to facilitate switching - both via a pin and a real time report element. The controller can then deny switching if in a state that it would be dangerous to do so.

**Reply by terjeio** [reply]
@stevempotter New build with changes to the keypad and MPG protocol up now. Next is to write documentation for how to use this. The keypad plugin can be used standalone with either I2C or UART serial communication, it can also be used in combination with the MPG mode option that allows taking/releasing full control either via a hardwired signal (I2C mode) or via a new real time command `0x8B` (UART mode). The hardwired signal uses the _Limit B_ input for now as it needs to be configured as an output during startup and as an input with pullup after. The Pro board allows redirecting this to the UART header TXE pin via a solder bridge.

Any further discussions related to this should take place at the grblHAL [core repo](https://github.com/grblHAL/core) or the [keypad plugin repo](https://github.com/grblHAL/Plugin_I2C_keypad) as it may be of interest to other grblHAL users.

**Reply by stevempotter** [reply]
An MPG toggle is good news. I look forward to the documentation so I can figure out how to use it.

> Any further discussions related to this should take place at the grblHAL core repo or the keypad plugin repo as it may be of interest to other grblHAL users.

OK, but first some teensy-specific questions. @phil-barrett please answer if you can.

>  The hardwired signal uses the Limit B input for now as it needs to be configured as an output during startup and as an input with pullup after. The Pro board allows redirecting this to the UART header TXE pin via a solder bridge.

What does Limit B have to do with I2C and MPG signals? Does "hardwired" mean the A and B outputs of the encoder are connected to pins on the Teensy?
I am using Limit B for my 2nd Y motor, for auto-squaring. Will using the hardwired signal for MPG prevent me from doing that?

What do you mean by "Pro board"? I have v.2.09 T41U5XBB. I don't see any mention of "pro board" in the readme for this repo or on Tindie.

**Reply by terjeio** [reply]
An [initial wiki page](https://github.com/grblHAL/core/wiki/MPG-and-DRO-interfaces) has now been put up. I will add a list of keypad command characters to the keypad plugin readme later.

> What does Limit B have to do with I2C and MPG signals?

It has nothing to do with I2C signalling, but was a convenient pin to use for the MPG on the Pro prototypeI have as it can be routed to the UART header via a small blob of solder.

Does "hardwired" mean the A and B outputs of the encoder are connected to pins on the Teensy?

No. It is possible to do so with the encoder plugin - but this is not used by the keypad plugin and MPG/DRO modes.

> Will using the hardwired signal for MPG prevent me from doing that?

No, one if the options is not to use a hardwired pin or another pin could be repurposed. See the wiki page for a bit more details.

Since grblHAL runs on a number of different controllers and boards with very different capabilities I have made several options available - some may support none, some all. You have to chose which to use to be within the limits of the controller you have.

**Reply by stevempotter** [reply]
Hooray! My wireless pendant is now working well and I made a YouTube video about it. See this post:
https://github.com/terjeio/ioSender/discussions/266
Thanks for all your help!
